### ISSUE TITLE
Negated types

### ISSUE THREAD

**Issue by zpdDG4gta8XKpMCd** [microsoft/TypeScript#4196]
Sometimes it is desired to forbid certain types from being considered.

Example: `JSON.stringify(function() {});` doesn't make sense and the chance is high it's written like this by mistake. With negated types we could eliminate a chance of it.

``` typescript
type Serializable = any ~ Function; // or simply ~ Function
declare interface JSON {
   stringify(value: Serializable): string;
}
```

Another example

``` typescript

export NonIdentifierExpression = ts.Expression ~ ts.Identifier

```

**Comment by aleksey-bykov** [comment]
Gotta think more about it.

**Comment by whitecolor** [comment]
It is interesting, it is possible to achieve somehow?

**Comment by siegebell** [comment]
@aleksey-bykov This would allow unions with a catch-all member, without overshadowing the types of the known members.

```typescript
interface A { type: "a", data: number }
interface B { type: "b", data: string }
interface Unknown { type: string ~"a"|"b", data: any }
type ABU = A | B | Unknown

var x : ABU = {type: "a", data: 5}
if(x.type === "a") {
  let y = x.data; // y should be inferred to be a number instead of any
} 
```

**Comment by SalathielGenese** [comment]
Following @mhegazy [comment](https://github.com/Microsoft/TypeScript/issues/18280#issuecomment-371892317) at #18280, I copy-paste this suggestion here...

---

I upvote `A & !B` especially the `!B` part... Over `Exclude` from #21847

**Comment by jack-williams** [comment]
Do negated types rely on completeness for type-checking?

**Comment by aleksey-bykov** [comment]
I think you should never put a question of what exactly `any - MyClass` is. I think negated types should be evaluated loosely and only when it comes to typechecks against certain types.

**Comment by jack-williams** [comment]
I agree. Is that not sort of like many types now, e.g. `number`. You never consider how to construct `number` because it's infinite: only test that a value belongs to it when you need it. What would be the (lazy) procedure for checking that `T` belongs to `A - B`, or `!B`?

**Comment by SalathielGenese** [comment]
```ts
// exclude all match of T from U
U & !T

// extract all match of T within U
T & U

// type to all but T
!T
```

> What would be the (lazy) procedure for checking that `T` belongs to `A - B`, or `!B`?

```ts
T extends (A & !B)
// or
T extends !B
```

**Comment by aleksey-bykov** [comment]
not until all type parameters (`A`, `B` and `T` in your example) are resolved to concrete types (`string`, `MyClass`, `null`) you can tell what `A - B` is

**Comment by aleksey-bykov** [comment]
so the procedure would be:
1. keep type expressions unevaluated until all type parameters are resolved
2. once all type parameters are known, replace them with the concrete types and see if the expression makes makes any sense

**Comment by jack-williams** [comment]
Understood! I guess my question is when you have some concrete types, say `A, B, C`, and you want to know if `A` is assignable to `B - C`, is it something like.
- if `A` is assignable to `B`
- and `A` is not assignable to `C`
- then `A` is assignable to `B - C`.

For `A` assignable to `!C` it would just be is `A` _not_ assignable to `B`.

Sorry if that's not clear!

**Comment by SalathielGenese** [comment]
> For A assignable to `!C` it would just be is `A` *not* assignable to `B`.

I think you meant **`A` *not* assignable to `C`**

**Comment by aleksey-bykov** [comment]
not sure if you can apply sort of a type algebra here, because it's unclear how assignability relates to negation 

what you can do is to build a concrete type out of `B - C` and name it `D` (provided both `B` and `C` are known) and then ask a question whether or now `A` is assignable to the concrete type `D`

my naive 5 cents

**Comment by aleksey-bykov** [comment]
question still stands what to do when B is too broad like `any`

**Comment by SalathielGenese** [comment]
I think a cleaner way to see `B - C` would be much like a type constraint rather a type by essence.

**Comment by SalathielGenese** [comment]
If by some logic it can be resolved to a type, that would be great, otherwise, it is just a type constraint

**Comment by zheeeng** [comment]
Expected progress on negating operate.
We already have `Exclude<T, U>` it is awesome that if the second type `U` is optional. We can easy implements `Not<T>` to exclude `T` from all types.
I also upvote using `~ T` or `unlike T` to constraints types.

**Comment by the1mills** [comment]
```js
export type NotUndefined = !undefined;
```

would be extremely useful IMO

**Comment by RyanCavanaugh** [comment]
`Exclude` gives the possibility to remove something from a union, and conditional types do some other good stuff.

For cases of actual subtype exclusion, the possibility of aliasing makes this idea sort of bonkers. "Animal but not Dog" doesn't make sense when you can alias a Dog via an Animal reference and no one can tell.

Anyway here's something that kinda works!
```ts
type Animal = { move: string; };
type Dog = Animal & { woof: string };

type ButNot<T, U> = T & { [K in Exclude<keyof U, keyof T>]?: never };

function getPet(allergic: ButNot<Animal, Dog>) { }

declare const a: Animal;
declare const d: Dog;
getPet(a); // OK
getPet(d); // Error
```

**Comment by RyanCavanaugh** [comment]
>  simply with a check that prevents people from committing the aliasing mistake you described?

What mistake?

**Comment by jpike88** [comment]
If 'Animal but not Dog' doesn't make sense, that is something TS can be aware of and disallow. But including something like ButNot into TS syntax I think is a good idea

**Comment by jpike88** [comment]
If `Dog = Animal & { woof:string }` then `Animal && !Dog` would be equivalent to `Animal & !(Animal & { woof:string })`, which would always evaluate to false.

But @RyanCavanaugh, if certain combinations are logically problematic, does TS not have the ability to know this and just throw an error on parse?

**Comment by jack-williams** [comment]
> If Dog = Animal & { woof:string } then Animal && !Dog would be equivalent to Animal & !(Animal & { woof:string }), which would always evaluate to false.

Can you not do this:

- `Animal && !Dog` = `Animal & !(Animal & { woof:string })`
- `Animal & !(Animal & { woof:string })` = `Animal & (!Animal | !{woof: string})` by DeMorgan
- `Animal & (!Animal | !{woof: string})` = `(Animal & !Animal) | (Animal & !{woof: string})` by union distribution
-  `(Animal & !Animal) | (Animal & !{woof: string})` = `never | (Animal & !{woof: string})` by contradiction
- `never | (Animal & !{woof: string})` = `Animal & !{woof: string}` by lattice minimum

That seems a reasonable type to me: anything that is an animal, but not with a woof field of type string.

**Comment by RyanCavanaugh** [comment]
What you're describing is just the `ButNot` type above, but with the `?` removed

**Comment by jack-williams** [comment]
I don't know if that reply was for me, but that was the intention of my post. There doesn't seem to be anything 'logically problematic' with `Animal && !Dog` or `ButNot<Animal, Dog>`.

**Comment by jvanbruegge** [comment]
I would suggest a backslash as syntax, as this is also what the set operation looks like.
But anyways, this would be nice to have, because AFAICT currently this is not expressable in typescript:
```ts
type Config = {
    foo: number;
    bar: number;
    [k in (string \ ("foo" | "bar"))]: string;
};
```

**Comment by isiahmeadows** [comment]
@RyanCavanaugh Could this be reconsidered, as an alternative to `awaited`/`promised` for typing promises? Here's how you could properly type native promises with this (this addresses [concerns listed here](https://github.com/Microsoft/TypeScript/issues/17621)):

```ts
// Note: `!T` means "all types but T"
// - `!unknown` = `never`
// - `!never` = `unknown`
interface PromiseLike<T, E = Error> {
	then(onResolve: (value: T) => any, onReject: (error: E) => any): any;
}

interface PromiseLikeCoerce<T extends !PromiseLike<any>, E = Error>
	extends PromiseLike<PromiseCoercible<T, E>, E> {}
type PromiseCoercible<T extends !PromiseLike<any>, E = Error> =
	T | PromiseLikeCoerce<T, E>;

interface PromiseConstructor {
	resolve<T extends !PromiseLike<any>>(value: PromiseCoercible<T>): Promise<T, never>;
	reject<E = Error>(value: E): Promise<never, E>;
	all<T extends Array<!PromiseLike<any>>, E = Error>(
		values: (
			{[I in keyof T]: PromiseCoercible<T[I]>} |
			Iterable<Await<T[number]>>
		)
	): Promise<T, E>;
	race<T extends !PromiseLike<any>, E = Error>(
		values: Iterable<PromiseCoercible<T>>
	): Promise<T, E>;
}

interface Promise<T extends !PromiseLike<any>, E = Error> {
	then(onResolve?: !Function, onReject?: !Function): Promise<T, E>;
	catch(onReject?: !Function): Promise<T, E>;
	then<U, F = E>(
		onResolve: (value: AwaitValue<T>) => AwaitValue<U, F>,
		onReject?: !Function,
	): Promise<U, E | F>;
	then<U, F = E>(
		onResolve: !Function,
		onReject: (error: E) => AwaitValue<U, F>,
	): Promise<T | U, F>;
	then<U, F = E>(
		onResolve: (value: AwaitValue<T>) => AwaitValue<U, F>,
		onReject: (error: E) => AwaitValue<U, F>,
	): Promise<U, F>;
	catch<U, F = E>(
		onReject: (error: E) => AwaitValue<U, F>,
	): Promise<T | U, F>;
	finally(onSettled: () => PromiseCoercible<any, any>): Promise<T, E>;
}
```

Note that `any` here has to be used since TS only has concrete types + `any` + `never`.

-----

Technically, you *could* type it with only conditional types...</summary>

...but it'd get very awkward very fast, and it'd be a bit counter-intuitive and a ***really*** ugly hack. It also doesn't assert the invariant of promise resolutions never containing promises.

[Link to playground](https://www.typescriptlang.org/play/index.html#src=interface%20PromiseLike%3CT%2C%20E%20%3D%20Error%3E%20%7B%0D%0A%09then(onResolve%3A%20(value%3A%20T)%20%3D%3E%20any%2C%20onReject%3A%20(error%3A%20E)%20%3D%3E%20any)%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Atype%20AwaitValue%3CT%2C%20E%20%3D%20Error%3E%20%3D%20%7B%0D%0A%20%20%20%200%3A%20T%3B%0D%0A%20%20%20%201%3A%20T%20extends%20PromiseLike%3Cinfer%20U%2C%20any%3E%20%3F%20AwaitValue%3CU%2C%20E%3E%20%3A%20never%3B%0D%0A%7D%5BT%20extends%20PromiseLike%3Cany%2C%20any%3E%20%3F%201%20%3A%200%5D%3B%0D%0A%0D%0Atype%20AwaitPromise%3CT%2C%20E%20%3D%20Error%3E%20%3D%20%7B%0D%0A%20%20%20%200%3A%20Promise%3CT%2C%20E%3E%3B%0D%0A%20%20%20%201%3A%20T%20extends%20PromiseLike%3Cinfer%20U%2C%20infer%20F%3E%20%3F%20AwaitPromise%3CU%2C%20E%20%7C%20F%3E%20%3A%20never%3B%0D%0A%7D%5BT%20extends%20PromiseLike%3Cany%2C%20any%3E%20%3F%201%20%3A%200%5D%3B%0D%0A%0D%0Ainterface%20PromiseLikeCoerce%3CT%2C%20E%20%3D%20Error%3E%0D%0A%09extends%20PromiseLike%3CPromiseCoercible%3CT%2C%20E%3E%2C%20E%3E%20%7B%7D%0D%0Atype%20PromiseCoercible%3CT%2C%20E%20%3D%20Error%3E%20%3D%20T%20%7C%20PromiseLikeCoerce%3CT%2C%20E%3E%3B%0D%0A%0D%0Ainterface%20PromiseConstructor%20%7B%0D%0A%09resolve%3CT%3E(value%3A%20T)%3A%20AwaitPromise%3CT%3E%3B%0D%0A%09reject%3CE%20%3D%20Error%3E(value%3A%20E)%3A%20Promise%3Cnever%2C%20E%3E%3B%0D%0A%09all%3CT%20extends%20%5Bany%2C%20any%5D%5B%5D%3E(%0D%0A%09%09values%3A%20(%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20Pick%3CT%2C%20Exclude%3Ckeyof%20T%2C%20number%3E%3E%20%26%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7B%5BI%20in%20number%20%26%20keyof%20T%5D%3A%20PromiseCoercible%3CT%5BI%5D%5B0%5D%2C%20T%5BI%5D%5B1%5D%3E%7D%20%7C%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20Iterable%3CPromiseCoercible%3CT%5Bnumber%5D%3E%3E%0D%0A%20%20%20%20%20%20%20%20)%0D%0A%09)%3A%20Promise%3C%0D%0A%20%20%20%20%20%20%20%20Pick%3CT%2C%20Exclude%3Ckeyof%20T%2C%20number%3E%3E%20%26%0D%0A%20%20%20%20%20%20%20%20%7B%5BI%20in%20number%20%26%20keyof%20T%5D%3A%20AwaitValue%3CT%5BI%5D%5B0%5D%2C%20T%5BI%5D%5B1%5D%3E%7D%2C%0D%0A%20%20%20%20%20%20%20%20T%5Bnumber%5D%5B1%5D%0D%0A%20%20%20%20%3E%3B%0D%0A%09race%3CT%2C%20E%20%3D%20Error%3E(values%3A%20Iterable%3CT%3E)%3A%20AwaitPromise%3CT%2C%20E%3E%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Promise%3CT%2C%20E%20%3D%20Error%3E%20%7B%0D%0A%09then%3CU%2C%20F%20%3D%20E%3E(%0D%0A%09%09onResolve%3A%20(value%3A%20AwaitValue%3CT%3E)%20%3D%3E%20AwaitValue%3CU%2C%20F%3E%2C%0D%0A%09%09onReject%3A%20(error%3A%20E)%20%3D%3E%20AwaitValue%3CU%2C%20F%3E%2C%0D%0A%09)%3A%20Promise%3CU%2C%20F%3E%3B%0D%0A%09catch%3CU%2C%20F%20%3D%20E%3E(%0D%0A%09%09onReject%3A%20(error%3A%20E)%20%3D%3E%20AwaitValue%3CU%2C%20F%3E%2C%0D%0A%09)%3A%20Promise%3CT%20%7C%20U%2C%20F%3E%3B%0D%0A%09finally(onSettled%3A%20()%20%3D%3E%20AwaitValue%3Cany%3E)%3A%20Promise%3CT%2C%20E%3E%3B%0D%0A%7D)

**Comment by webhacking** [comment]
Checkout 3.5
https://github.com/Microsoft/TypeScript/issues/30555

**Comment by isiahmeadows** [comment]
Could this be re-opened in light of #29317?

**Comment by kelsny** [comment]
I want this feature in some way or another, maybe even as a marker type like `ThisType`.

I have a "language" for matching tree-like structures:

```
{
    type: "NOT_NUMBER";
    value: not number;
}
```

In this language of mine, I support `not`, which matches all types but its operand. So `not string and not number` would disallow strings and numbers.

I've got most of this excellently (over engineered to hell) typed:

```ts
fantasyLibrary.match(compiledExpression, myTree, (match) => {
    // match is of type { type: "NOT_NUMBER"; value: unknown; }
});
```

The `match` parameter is as close as possible to what the original source expression describes. 
However, because TypeScript doesn't have any way of representing the negation of types, I am forced to type anything that uses `not` as `unknown`, which isn't as great as I would've liked it to be.

If this was a thing in TypeScript, I would be able to provide even more accurate types in the callback of my library function.

**Comment by Pyrolistical** [comment]
I think the most obvious use-case is being able to define a non-empty string type:

```ts
type NonEmptyString = string & !''
```

`Exclude<string, ''>` doesn't work because `Exclude` is only for union types.

**Comment by HolgerJeromin** [comment]
@Pyrolistical 
Typescript is mostly about types of variables (and other things) and not about its content..
ref `NaN` discussion: https://github.com/microsoft/TypeScript/issues/28682#issuecomment-707142417

**Comment by Pyrolistical** [comment]
@HolgerJeromin but typescript does allow string literal types. How is this different?

**Comment by thw0rted** [comment]
Not just string literals -- I can think of a number of places where I would find `Exclude<number, 0>` helpful. Similar to the post above, I'd like to just have some way to say "Type X, but not falsy".

**Comment by HolgerJeromin** [comment]
> but typescript does allow string literal types. How is this different?

They are kind of enums. 
So with them we have a positive list (this "few" values), but right now no negative list (all but not these values).
The type `"auto" | string` gets `string`.

**Comment by darrylnoakes** [comment]
Would this work with template literal types? E.g. to exclude certain substrings from a string. (I'm assuming it would.)
If so, I want this even more than before!

```ts
// Error: Type '"Hello, foo!"' is not assignable to type 'not `${string}foo${string}`'. ts(2322)
const withoutFoo: not `${string}foo${string}` = "Hello, foo!";

// And with generics:
type NoSubstring<substr extends string> = not `${string}${substr}${string}`;
const withoutFoo: NoSubstring<"foo"> = "Hello, foo!";
```

**Comment by douglasg14b** [comment]
"No Recent Activity"

Yet another dead link in a bunch of other issues >_>

Issue fragmentation and lack of coherency is getting to be a problem on large, active, repos

**Comment by KenjiTakahashi** [comment]
I wonder how far are we from people creating bots that create artificial activity so that bots detecting no activity do not close relevant issues?

**Comment by truemogician** [comment]
Being mentioned in other issues or pull requests should also be considered as "activity".

**Comment by RyanCavanaugh** [comment]
Apologies, this one was mis-tagged. Negated types are still "on the table" so to speak.

**Comment by dead-claudia** [comment]
@RyanCavanaugh Might be worth creating a label to force-keep issues like this open.

**Comment by jakebailey** [comment]
Note the label change above; we have a set of labels which imply an eventual close (but were inconstantly done automatically) and this issue no longer has one.

**Comment by mitchell-merry** [comment]
I'd like to throw my use-case into the mix ([playground](https://tsplay.dev/wQJqZN)):

```ts
type MyType = string | 'some-specific-string';

type IsSpecific<T extends MyType> = T extends 'some-specific-string' ? true : false;
type ShouldBeTrue = IsSpecific<'some-specific-string'>;
//   ^?
type ShouldBeFalse = IsSpecific<'unspecific'>;
//   ^?

type MyMappedType = {
  [K in MyType]: IsSpecific<K>
};

const current: MyMappedType = {
  'unspecific': false, // OK
  'some-specific-string': false, // should error
};
```
Here, I want to allow arbitrary keys into a dictionary, but if a key matches a particular string sub-type then the value should be different. This doesn't work, since `some-specific-string` gets resolved to the type `string` and therefore has a value of false.

I think if the `not` type was a thing, then the following would do what I need:
```ts
type MyType = (string & not 'some-specific-string') | 'some-specific-string';

type MyMappedType = {
  [K in MyType]: IsSpecific<K>
};

const a: MyMappedType = {
  'unspecific': false,
  'some-specific-string': true,
};
```

This way, `MyType` still matches any `string`, but it would prevent the arbitrary string type from swallowing `some-specific-string`. I'm not sure if there's a way to solve this in TypeScript as it stands.

+1 to this feature request.

**Comment by BetaZhang** [comment]
`
type Falsy = false | 0 | "" | null | undefined;type Truthy = !Falsy;
`

**Comment by BetaZhang** [comment]
`type Falsy = false | 0 | "" | null | undefined; `
`type Truthy = !Falsy;`

**Comment by ljharb** [comment]
you forgot 0n and NaN and document.all :-)

**Comment by thw0rted** [comment]
If you, like any reasonable person, said "WTF" out loud on reading the above comment, feel free to [marvel at the bad decisions we sometimes make](https://stackoverflow.com/questions/10350142/why-is-document-all-falsy) in pursuit of backward-compatibility.

**Comment by BetaZhang** [comment]
> you forgot 0n and NaN and document.all :-)

Thanks for your correction.

**Comment by yamiteru** [comment]
I'd like to add my use-case. I'm working on a data validation library and I want to have an operation `not` which takes as a parameter another operation (for example `literal(null)`) and it should return a type that's anything except the output type of the input operation so in this case something like `!null`.

The types can also be custom so I cannot use `Exclude` in this case because I don't know all of the custom types beforehand so I cannot create a union of all those possible types.

**Comment by MKRhere** [comment]
Telegram Bot API recently added `type MaybeAccessibleMessage = Message | InaccessibleMessage`, of which the specified way to check is `message.date === 0` (InaccessibleMessage).

As the maintainers of very popular Bot API libs for TypeScript @telegraf and @grammyjs, we take effort to model [these](https://github.com/telegraf/types) [types](https://github.com/grammyjs/types) for TypeScript as accurately as possible. Negated types would be super useful in this among many other usecases, to define `Message["date"]` as `number ~ 0`, so you can discriminate based on it.

I just hope we can restart a conversation on this long pending issue. Are there obvious design reasons this is not viable today?

**Comment by jorenbroekema** [comment]
I'd love negated types for the following use case:
https://www.typescriptlang.org/play/?#code/LAKFFMA8AcHsCcAuACAlgO0eeAzAhgMbjIDyARgFbIDeoyyOssA-AFzIDOi8GA5gNx1kAbQDW4AJ7suPdLwC67chUEgAvqCA

```ts
export interface Obj {
  foo?: string; // this errors, because Property 'foo' of type 'string | undefined' is not assignable to 'string' index type 'Obj'
  [key: string]: Obj;
}
```

I'm solving it now by doing this:

```ts
export interface Obj {
  foo?: string;
  [key: string]: Obj | string | undefined;
}
```
But that's a little bit weird because I know that when the key is `"foo"`, it's `string | undefined` and when the key is NOT `"foo"`, it means it is a nested version of itself.

So preferably I would do this:

```ts
export interface Obj {
  foo?: string;
  [key: string & not keyof Obj]: Obj;
}
```

But I'm not really sure if that use case would work because technically `[key: string & not keyof Obj]: Obj;` is a part of `keyof Obj`?

**Comment by kaleidawave** [comment]
I don't think this would be too hard to implement. T being a subtype of 'not U' is equivalent to T being disjoint from U (aka no common members). [This disjoint logic is already done for the checking of JS equality](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAHjAvDATAbgFABMCmwA2AhgE44z46wCeAXDNMQJZgDmmGjAZjABQKICYVAJQwA3gF8gA), so could probably could be reused in the subtyping logic. 

This would also solve some of the problems with `Exclude`. `Exclude` normally works using the extends distribution mechanism, but this only happens for union types and doesn't distribute across all floating point types. With `p: Not<2>` the first two checks fail disjoint checks and therefore wouldn't be allowed to be assigned to the parameter.

```ts
function func<const T>(p: Exclude<T, 2>) {}

func(2) // <- errors
func(2 as number) // <- no error (when there should be)
func(6) // <- works as intended
```
###### https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABKSAeCCDOVEBUB8AFAA4BciAogB4QA2IAJgKaq4A0iATPgJSIDeAXwBQwlBEKceY8BIBs08ZMQBDTIjAgAtgCMmAJx5A

**Comment by onx2** [comment]
I may have been stated before in another ticket but the use case I have for this feature is containing a string so:

**Pseudo code**
```ts
// Where `T` is a union of string literals
type AnyStringExcept<T> = string & !T;
type NotAllowed = "a" | "b";

AllowedStrings = string | AnyStringExcept<NotAllowed>;
```

As its been mentioned before, this behavior can exist in a function but not at a "type-only level". 

The project specific use-case for me is constraining the keys that are pressed for a keybinding to only valid characters.


```ts
// Every possible universal key across all keyboards
export type UniversalKey = 
  // Control Keys
  | "BACKSPACE" | "TAB" | "ENTER" | "ESCAPE" | "DELETE"
  // Modifier Keys
  | "SHIFT" | "CONTROL" | "ALT" | "META"
  // Arrow Keys
  | "ARROWLEFT" | "ARROWUP" | "ARROWRIGHT" | "ARROWDOWN"
  // Number Keys
  | "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
  // Numpad Keys
  | "NUMPAD0" | "NUMPAD1" | "NUMPAD2" | "NUMPAD3" | "NUMPAD4"
  | "NUMPAD5" | "NUMPAD6" | "NUMPAD7" | "NUMPAD8" | "NUMPAD9"
  | "NUMPADMULTIPLY" | "NUMPADADD" | "NUMPADSUBTRACT" | "NUMPADDECIMAL" | "NUMPADDIVIDE"
  // Function Keys
  | "F1" | "F2" | "F3" | "F4" | "F5" | "F6" 
  | "F7" | "F8" | "F9" | "F10" | "F11" | "F12"
  // Navigation Keys
  | "HOME" | "END" | "PAGEUP" | "PAGEDOWN"
  // Other Special Keys
  | "CAPSLOCK" | "INSERT" | "PAUSE" | "PRINTSCREEN" | "SCROLLLOCK";

// a constrained list of universal keys that are acceptable for keybindings
export type ValidUniversalKey = Extract<UniversalKey,
  // Control Keys
  "BACKSPACE" | "TAB" | "ENTER" | "ESCAPE" | "DELETE"
  // Modifier Keys
  | "SHIFT" | "CONTROL" | "ALT" | "META"
  // Arrow Keys
  | "ARROWLEFT" | "ARROWUP" | "ARROWRIGHT" | "ARROWDOWN"
  // Number Keys
  | "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
  // Numpad Keys
  | "NUMPAD0" | "NUMPAD1" | "NUMPAD2" | "NUMPAD3" | "NUMPAD4"
  | "NUMPAD5" | "NUMPAD6" | "NUMPAD7" | "NUMPAD8" | "NUMPAD9">;

export type InvalidUniversalKey = Exclude<UniversalKey, ValidUniversalKey>;

export type ValidShortcutKey = (ValidUniversalKey | ValidUniqueKey) & AnyStringExcept<InvalidUniversalKey>;
```

Having this would prevent me from loosening the type def or defining a massive character list (string literals).

**Comment by adueck** [comment]
This would be really nice to have because I could use combinators like AND, OR, NOT, for type predicates.

```ts
/**
 * a type predicate AND combinator
 */
export function andTp<A, B extends A, C extends A>(
  f: (x: A) => x is B,
  g: (x: A) => x is C
): (x: A) => x is B & C {
  return (x: A) => f(x) && g(x);
}
```

```ts
/**
 * a type predicate OR combinator
 */
export function orTp<A, B extends A, C extends A>(
  f: (x: A) => x is B,
  g: (x: A) => x is C
): (x: A) => x is B | C {
  return (x: A) => f(x) || g(x);
}
```

This works great and I'm able to do point-free combinations of type predicates. But I can't see how I can do NEGATION of type predicates without negated types.

```ts
// I wish I could do this...
export function notTp<A>(
  f: (x: A) => x is A,
): (x: A) => x is not A {
  return (x: A) => !f(x);
}
```

Then I could nicely compose my type predicates and do things like this

```ts
array.filter(andTp(isFoo, notTp(isBar)))
```

**Comment by adrianstephens** [comment]
I'm trying to determine the current state of this and related suggestions (going back over 10 years). It seems like it's in popular demand, and the lack of such a feature makes some valid JS unrepresentable (if you want it typed).

Is it just not as high a priority as other features (and if so, how are the priorities determined)?

Is it a philosophically problematic addition for some reason?

Is the implementation considered too hard or risky vs perceived utility?

If someone implemented type subtraction or negation (and it didn't break anything!) would that influence the decision to include the feature? Would a PR be welcome (asking for a friend ;) )?

**Comment by lucasbasquerotto** [comment]
@adrianstephens My guess is that implementing negated types in a duck-typed language like TypeScript would be quite challenging. Since TypeScript allows structural typing, an object of one type can often be used as another type if they share the same attributes—even if their meanings are entirely different.

```ts
// A and B are distinct interfaces but have identical attributes
const fn = (a: A, callback: (notA: ~A) => void) => {
    const b: B = a; // Structurally identical, so assignment is allowed
    callback(b); // This is valid
    callback(a); // This is an error
};

// The issue becomes even more problematic when implicit type conversions occur
const fn2 = (b: B, callback: (notA: ~A) => void) => {
    callback(b); // Accepted
};

const a: A = getA();
fn2(a, (notA: ~A) => console.log('not A', notA)); // Unexpectedly valid

// Ambiguity arises with inheritance and type intersections
interface C extends A, B {}
// Alternatively, C could be defined as: type C = A & B

function getC(): C { ... }
const c1 = getC();
fn2(c1, (notA: ~A) => console.log('not A', notA)); // Error
const c2: B = getC();
fn2(c2, (notA: ~A) => console.log('not A', notA)); // Allowed
```

These cases could pose challenges for those looking to implement negated types, depending on how strictly they need to be enforced. The validity of ~A depends not only on the structure of A but also on how TypeScript resolves assignments and type relations dynamically.

One potential solution would be introducing nominal types (see: https://github.com/Microsoft/Typescript/issues/202). If nominal typing were supported, negated types might be more feasible—at least for explicitly nominal types—since they wouldn't rely purely on structural compatibility.

**Comment by adrianstephens** [comment]
@lucasbasquerotto Without thinking about it too hard, it seems to me that if you have a function in your transpiler that determines if one type is assignable to another (say), you can just negate the result for negated types. In other words, just think of the type checker as working on abstract things that obey boolean algebra.

I think type subtraction is a bit clearer conceptually, so that's what I tried to implement; but negation would probably've been simpler.

**Comment by Rudxain** [comment]
My 2 cents: Even though Rust and TS have **very** different type-systems, [discussions about negative trait-bounds](https://github.com/rust-lang/rust/issues/42721) (and specialization) could help reach a consensus about this

**Comment by yuhr** [comment]
Control flow analysis already seems to do this for `else` clauses, doesn't it?

**Comment by adrianstephens** [comment]
The thing is, all these nuances have already been discussed in this very long thread - probably multiple times! I don't think this feature is pending a specification. Which brings me back to my original question - what *is* the status, and what can be done to push it forward?

**Comment by amine-err** [comment]
I would also love this for multiple projects I'm working on, while it isn't blocking as I already have working implementations, but it would make working with typescript way easier and cleaner.

I'm fairly new to typescript, so probably no one will need this like I did, but if anyone is interested, here are some examples of how I implemented negation for my projects.

```typescript
/** Given a type `T`, it returns `T` if it extends `A` AND does not extend `N`, if not it returns never */
type Not<N, T, A = any> = [T] extends [N] ? never : (
  [T] extends [A] ? T : never
);

// Non empty strings
type NonEmptyString<T> = Not<'', T, string>;
function takesNonEmptyString<T>(v: NonEmptyString<T>) {};

takesNonEmptyString(''); // gives a type error as it is an empty string
takesNonEmptyString('not empty'); // works as expected
takesNonEmptyString(20); // gives a type error as it isn't a string

// Non zero number
type NonZeroNumber<T> = Not<0, T, number>;
function takesNonZeroNumber<T>(v: NonZeroNumber<T>) {}

takesNonZeroNumber(0); // gives a type error as it is 0
takesNonZeroNumber(20); // works as expected
takesNonZeroNumber('str'); // gives a type error as it isn't a number

// Any string but a set
type AnyStringBut<T> = Not<'a' | 'b', T, string>;
function takesAnyStringBut<T>(v: AnyStringBut<T>) {};

takesAnyStringBut('a'); // gives a type error as it is 'a'
takesAnyStringBut('b'); // gives a type error as it is 'b'
takesAnyStringBut('c'); // works as expected as any other string
takesAnyStringBut(20); // gives an error as it isn't a string

// Infer value using generic function
function not<N, A = any>() {
  return { infer<T>(v: Not<N, T, A>) { return v } }
}

const nonEmptyString = not<'', string>().infer(''); // gives a type error as it is an empty string
const nonZeroNumber = not<0, number>().infer(0); // gives a type error as it is 0
const anyButFunction = not<Function>().infer(()=>{}) // gives  a type error as it is a function
const anyStringBut = not<'a' | 'b', string>().infer('b') // gives a type error as it is 'b'
```

This correctly gives type errors if the value is not allowed:

<img width="1480" height="821" alt="Image" src="https://github.com/user-attachments/assets/18b40c70-321b-42b0-81f7-033651cf7dd9" />

**Comment by yuri-kiss** [comment]
> I would also love this for multiple projects I'm working on, while it isn't blocking as I already have working implementations, but it would make working with typescript way easier and cleaner.
> 
> I'm fairly new to typescript, so probably no one will need this like I did, but if anyone is interested, here are some examples of how I implemented negation for my projects.
> 
> type Not<N, T, A = any> = [T] extends [N] ? never : (
>   [T] extends [A] ? T : never
> );
> 
> // Non empty strings
> type NonEmptyString<T> = Not<'', T, string>;
> function takesNonEmptyString<T>(v: NonEmptyString<T>) {};
> 
> takesNonEmptyString(''); // gives a type error as it is an empty string
> takesNonEmptyString('not empty'); // works as expected
> takesNonEmptyString(20); // gives a type error as it isn't a string
> 
> // Non zero number
> type NonZeroNumber<T> = Not<0, T, number>;
> function takesNonZeroNumber<T>(v: NonZeroNumber<T>) {}
> 
> takesNonZeroNumber(0); // gives a type error as it is 0
> takesNonZeroNumber(20); // works as expected
> takesNonZeroNumber('str'); // gives a type error as it isn't a number
> 
> // Any string but a set
> type AnyStringBut<T> = Not<'a' | 'b', T, string>;
> function takesAnyStringBut<T>(v: AnyStringBut<T>) {};
> 
> takesAnyStringBut('a'); // gives a type error as it is 'a'
> takesAnyStringBut('b'); // gives a type error as it is 'b'
> takesAnyStringBut('c'); // works as expected as any other string
> takesAnyStringBut(20); // gives an error as it isn't a string
> 
> // Infer value using generic function
> function not<N, A = any>() {
>   return { infer<T>(v: Not<N, T, A>) { return v } }
> }
> 
> const nonEmptyString = not<'', string>().infer(''); // gives a type error as it is an empty string
> const nonZeroNumber = not<0, number>().infer(0); // gives a type error as it is 0
> const anyButFunction = not<Function>().infer(()=>{}) // gives  a type error as it is a function
> const anyStringBut = not<'a' | 'b', string>().infer('b') // gives a type error as it is 'b'
> This correctly gives type errors if the value is not allowed:
> 
> <img alt="Image" width="1480" height="821" src="https://private-user-images.githubusercontent.com/118001268/528919820-18b40c70-321b-42b0-81f7-033651cf7dd9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY1ODg0NDksIm5iZiI6MTc2NjU4ODE0OSwicGF0aCI6Ii8xMTgwMDEyNjgvNTI4OTE5ODIwLTE4YjQwYzcwLTMyMWItNDJiMC04MWY3LTAzMzY1MWNmN2RkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIyNFQxNDU1NDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZDA0NDFmNTlmNjE4MzllZGI2Y2M3YzEyNmZlY2IxM2QwNzE3ZDk3YjlhZGE2ZjE5ZDAwZTRiNmVkZjMzYTc2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.F4XhcLVcc-jL_5PsddKT6E4OjuVHknwLBR2G3CyOGis">

is there a way to do this for like `Infinity | -Infinity | NaN`, im trying to make a "FiniteNumber" type

**Comment by amine-err** [comment]
> is there a way to do this for like `Infinity | -Infinity | NaN`, im trying to make a "FiniteNumber" type

@yuri-kiss I don't think there is a straightforward way to do this at the type level, as `Infinity` and `NaN` are not types but values, both are typed as numbers and are not considered as literal types by typescript, so there is no way to distinguish between them and a finite number at the type level. So this won't be fixed even if negated types are added, unless typescript recognizes `Infinity` and `NaN` as types.

Tho, after thinking about it and searching for a bit, I did find a way to implement this, tho I prefer to not post here, as it isn't directly related to this issue, and I don't want to clutter this discussion.

As such, I answered your question and explained how I did it in this [page](https://github.com/amine-err/typescript-helpers/wiki/TS-FiniteNumber-Helper-type).

**Comment by Rudxain** [comment]
> "FiniteNumber"

[this almost works](https://github.com/sindresorhus/type-fest/blob/fe88ad8e4cee79dcb64f1c26f648319bfbcc4308/source/numeric.d.ts#L34-L51) (more info [here](https://github.com/microsoft/TypeScript/issues/31752#issuecomment-3693409406))

**Comment by yuri-kiss** [comment]
> > is there a way to do this for like `Infinity | -Infinity | NaN`, im trying to make a "FiniteNumber" type
> 
> [@yuri-kiss](https://github.com/yuri-kiss) I don't think there is a straightforward way to do this at the type level, as `Infinity` and `NaN` are not types but values, both are typed as numbers and are not considered as literal types by typescript, so there is no way to distinguish between them and a finite number at the type level. So this won't be fixed even if negated types are added, unless typescript recognizes `Infinity` and `NaN` as types.
> 
> Tho, after thinking about it and searching for a bit, I did find a way to implement this, tho I prefer to not post here, as it isn't directly related to this issue, and I don't want to clutter this discussion.
> 
> As such, I answered your question and explained how I did it in this [page](https://github.com/amine-err/typescript-helpers/wiki/TS-FiniteNumber-Helper-type).

thank you both so much
